consensus/src/dag/dag_state_sync.rs

```diff
@@ -0,0 +1,180 @@
+    pub async fn sync_to_highest_commit_cert(&self, ledger_info: &LedgerInfoWithSignatures) {
+        let dag_reader = self.dag_store.read();
+
+        // if the anchor exists between ledger info round and highest ordered round
+        // Note: ledger info round <= highest ordered round
+        if dag_reader.highest_committed_round().unwrap_or_default()
+            < ledger_info.commit_info().round()
+            && dag_reader.exists_by_round_digest(
+                ledger_info.commit_info().round(),
+                ledger_info.ledger_info().consensus_data_hash(),
+            )
+            && dag_reader.highest_ordered_round().unwrap_or_default()
+                >= ledger_info.commit_info().round()
+        {
+            self.network.send_commit_proof(ledger_info.clone()).await
+        }
+    }
+
+    /// Check if we're far away from this ledger info and need to sync.
+    /// This ensures that the block referred by the ledger info is not in buffer manager.
+    pub fn need_sync_for_ledger_info(&self, li: &LedgerInfoWithSignatures) -> bool {
+        let dag_reader = self.dag_store.read();
+        (dag_reader.highest_ordered_round().unwrap_or_default() < li.commit_info().round()
+            && !dag_reader.exists_by_round_digest(
+                li.commit_info().round(),
+                li.ledger_info().consensus_data_hash(),
+            ))
+            || dag_reader.highest_committed_round().unwrap_or_default() + 2 * DAG_WINDOW
+                < li.commit_info().round()
+    }
+
+    pub async fn try_sync_to_highest_ordered_anchor(
+        &self,
+        node: &CertifiedNodeWithLedgerInfo,
+    ) -> anyhow::Result<Option<Arc<RwLock<Dag>>>> {
+        // Check whether to actually sync
+        let commit_li = node.ledger_info();
+        if !self.need_sync_for_ledger_info(commit_li) {
+            return Ok(None);
+        }
+
+        let dag_fetcher = Arc::new(DagFetcher::new(
+            self.epoch_state.clone(),
+            self.network.clone(),
+            self.time_service.clone(),
+        ));
+
+        self.sync_to_highest_ordered_anchor(node, dag_fetcher).await
+    }
+
+    /// Note: Assumes that the sync checks have been done
+    pub async fn sync_to_highest_ordered_anchor(
+        &self,
+        node: &CertifiedNodeWithLedgerInfo,
+        dag_fetcher: Arc<impl TDagFetcher>,
+    ) -> anyhow::Result<Option<Arc<RwLock<Dag>>>> {
+        let commit_li = node.ledger_info();
+
+        // TODO: there is a case where DAG fetches missing nodes in window and a crash happens and when we restart,
+        // we end up with a gap between the DAG and we need to be smart enough to clean up the DAG before the gap.
+
+        // Create a new DAG store and Fetch blocks
+        let target_round = node.round();
+        let start_round = commit_li.commit_info().round().saturating_sub(DAG_WINDOW);
+        let sync_dag_store = Arc::new(RwLock::new(Dag::new(
```

Comment:
we probably need a different constructor for this, this'll re-construct the current dag from storage